Add media video preload article

beaufortfrancois opened the pull request:
This is my WIP (not ready at all) article for video preload. Let's use it to get feedback.

Live preview at https://pr-4744-dot-web-central.appspot.com/web/fundamentals/media/fast-playback-with-video-preload

jpmedley commented on src/content/en/fundamentals/media/video-preload.md:
```diff
@@ -0,0 +1,393 @@
+
+### Preload full video
+
+Here's how to use it to preload a full video  on your website so that
```
You can delete 'to use it' without clouding the meaning.

beaufortfrancois replied:
Done.

jpmedley commented on src/content/en/fundamentals/media/video-preload.md:
```diff
@@ -0,0 +1,393 @@
+Here's how to use it to preload a full video  on your website so that
+when your JavaScript asks to fetch video content, it is read from disk cache as
+the resource may have already been cached by the browser. If preload request
+didn't finish yet, a regular network fetch will happen. I'd recommend using
```
'hasn't finished' not 'didn't finish'.

beaufortfrancois replied:
Done.

jpmedley commented on src/content/en/fundamentals/media/video-preload.md:
```diff
@@ -0,0 +1,393 @@
+when your JavaScript asks to fetch video content, it is read from disk cache as
+the resource may have already been cached by the browser. If preload request
+didn't finish yet, a regular network fetch will happen. I'd recommend using
+this only for audio songs or short video clips.
```
change 'audio songs' to 'audio files'.

beaufortfrancois replied:
Done.

jpmedley commented on src/content/en/fundamentals/media/video-preload.md:
````diff
@@ -0,0 +1,399 @@
+</script>
+```
+
+As the preloaded resource is going to be consumed by a video element in the
````
Change 'As' to 'Because'.

beaufortfrancois replied:
Done.

jpmedley commented on src/content/en/fundamentals/media/video-preload.md:
```diff
@@ -0,0 +1,399 @@
+
+### Preload first segment
+
+Here's another example below of how to preload the first segment of a video
```
Either:

'Here's another example of how...'
_or_
'The example below shows how...'

I prefer the second.

beaufortfrancois replied:
Second then ;) Done.

jpmedley commented on src/content/en/fundamentals/media/video-preload.md:
```diff
@@ -0,0 +1,399 @@
+not familiar with the MSE Javascript API, please read [MSE basics].
+
+For the sake of simplicity, let's assume the entire video has been split into
+smaller files like "file_1.webm", "file_2.webm", "file_3.webm", etc. We'll use
```
If I'm using this technique, must I use video segments, or is there a way to do it with range requests?

beaufortfrancois replied:
I believe HTTP Range requests can't be called from `<link rel=preload href=...`

jpmedley commented on src/content/en/fundamentals/media/video-preload.md:
```diff
@@ -0,0 +1,399 @@
+
+## Manual Buffering
+
+Before we dive into the Cache API and Service Workers, let's see how to
```
lower case the word service workers.

beaufortfrancois replied:
Done.

jpmedley commented on src/content/en/fundamentals/media/video-preload.md:
```diff
@@ -0,0 +1,399 @@
+Before we dive into the Cache API and Service Workers, let's see how to
+manually buffer a video with MSE. The example below assumes that your web
+server supports HTTP range requests. Note that this would be pretty similar
+with segment files.
```
'similar to' not 'similar with'

beaufortfrancois replied:
Done.

beaufortfrancois commented on src/content/en/fundamentals/media/video-preload.md:
```diff
@@ -0,0 +1,393 @@
+{# wf_published_on: 2017-06-21 #}
+{# wf_updated_on: 2017-06-21 #}
+
+# Video Preload = Faster Playback {: .page-title }
```
Done.

beaufortfrancois commented on src/content/en/fundamentals/media/video-preload.md:
```diff
@@ -0,0 +1,393 @@
+ignore the `preload` attribute. At the time of writing, here are some rules
+applied in Chrome:
+
+- When [Data Saver]{: .external} is enabled, Chrome forces `preload` value to
```
Done.

beaufortfrancois commented on src/content/en/fundamentals/media/video-preload.md:
```diff
@@ -0,0 +1,393 @@
+If the video source is an unique file hosted on a web server, you may want to
+use the video `preload` attribute to provide a hint to the browser as to [how
+much information or content to preload]. In other words, [Media Source
+Extensions (MSE)] is not compatible with `preload`.
```
@mounirlamouri can confirm but I believe `preload` could work as a fallback for MSE.

jpmedley commented:
@beaufortfrancois Is this ready to go?

beaufortfrancois commented:
@jpmedley I've asked @johnpallett to have a look. As soon as it looks good to him, I'll definitely ping you. Thank you!

mounirlamouri commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
````diff
@@ -0,0 +1,492 @@
+```
+
+Note that if were to use HTTP Range request, I'd have to recreate manually a
+`Response` object as the Cache API doesn't support Range responses yet. Be
````
Maybe provide a link to the spec issue?

beaufortfrancois replied:
Done.

beaufortfrancois commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,492 @@
+  <source src="https://storage.googleapis.com/webfundamentals-assets/videos/video-preload-hero.webm#t=0.7"
+          type="video/webm">
+  <source src="https://storage.googleapis.com/webfundamentals-assets/videos/video-preload-hero.mp4#t=1"
+          type="video/mp4">
```
@johnpallett suggested I should not use YouTube there. Yes, I've manually modified code to make it look slower to load.

johnpallett replied:
Yes, please not YouTube

beaufortfrancois commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,492 @@
+
+Resource fetching will start only when initial HTML document has been
+completely loaded and parsed (eg. `DOMContentLoaded` event) while the very
+different `Load` event will be fired when resource has actually been fetched.
```
Done.

beaufortfrancois commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,492 @@
+</figure>
+
+Setting `preload` value to `metadata` indicates that the user is not expected
+to need the video, but that fetching its metadata (dimensions, first frame,
```
Done.
Please let me know when this is not the case anymore.

beaufortfrancois commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
````diff
@@ -0,0 +1,492 @@
+```
+
+With `preload` value set to `auto`, we're saying that downloading the entire
+video right away is considered desirable.
````
[Spec says](https://dev.w3.org/html5/spec-preview/media-elements.html#attr-media-preload) "Hints to the user agent that the user agent can put the user's needs first without risk to the server, up to and including optimistically downloading the entire resource."

Implementation differs: https://www.stevesouders.com/blog/2013/04/12/html5-video-preload/

What shall I say?

beaufortfrancois commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,492 @@
+
+There are some caveats though. As this is just a hint, browser may completely
+ignore the `preload` attribute. At the time of writing, here are some rules
+applied in Chrome:
```
The `/web/media` section is a living document and *will* be updated hence why I think it's good to give details when we can. I'll address your feedback below.

dalecurtis replied:
I'd also add that while the browser may ignore preload=auto, I don't think anyone ignores preload=metadata or preload=none.

beaufortfrancois commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,492 @@
+applied in Chrome:
+
+- When [Data Saver] is enabled, Chrome forces `preload` value to `none` if
+  video URL starts with "http" or "https".
```
Done.

beaufortfrancois commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,492 @@
+
+- When [Data Saver] is enabled, Chrome forces `preload` value to `none` if
+  video URL starts with "http" or "https".
+- In Android 4.3, Chrome forces `preload` value to `none`.
```
I'm not sure I understand your statement. Can you clarify?

beaufortfrancois commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,492 @@
+  video URL starts with "http" or "https".
+- In Android 4.3, Chrome forces `preload` value to `none`.
+- On a cellular connection (2G, 3G, and 4G), Chrome forces `preload` value to
+  `metadata`.
```
What is the bug to follow?

beaufortfrancois commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
````diff
@@ -0,0 +1,492 @@
+</script>
+```
+
+Note: I'd recommend using this only for audio files or short video clips.
````
Done.

beaufortfrancois commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,492 @@
+
+### Battery awareness
+
+I'd suggest you take into account the battery level of user's device before
```
Done.

beaufortfrancois commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,492 @@
+
+Now, what if I want to speculatively pre-load some media content without
+knowing which piece of media user will eventually pick. If user is on a web
+page that contains 10 videos, we probably have enough memory to fetch 1 segment
```
Done.

beaufortfrancois commented:
@mounirlamouri I'll add intro and conclusion this week.

dalecurtis commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,559 @@
+    </tr>
+    <tr>
+      <td rowspan=3 style="white-space: nowrap">
+<a href="#link_preload">Link preload</a>
```
I'm unfamiliar with the workings of this; so you might want to run it by the preload team too.

beaufortfrancois replied:
Noted.

dalecurtis commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,559 @@
+    </tr>
+    <tr>
+      <td>
+Doesn't work with MSE
```
Not really a "con" since the page is responsible for buffering in MSE.

beaufortfrancois replied:
Shall I rename "con" to "limitations" then to be more accurate?

beaufortfrancois replied:
I've tried something else... what do you think of https://github.com/google/WebFundamentals/pull/4744/commits/02439c38d3d4115bedab69952316fe7935bfe000

dalecurtis commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
````diff
@@ -0,0 +1,559 @@
+</script>
+```
+
+With `preload` value set to `auto`, we're saying th at downloading the entire
````
s/th at/that/

dalecurtis replied:
I also wouldn't use the words "the entire video". I don't think any browser does that right away, they only cache enough such that canplaythrough is satisfied.

beaufortfrancois replied:
> I'd also add that while the browser may ignore preload=auto, I don't think anyone ignores preload=metadata or preload=none.

Browser ignores `preload=metadata` in several conditions: Chrome Data Saver and Android 4.3.

> s/th at/that/
 
Fixed.

> I also wouldn't use the words "the entire video". I don't think any browser does that right away, they only cache enough such that canplaythrough is satisfied.

Fixed.

dalecurtis commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,559 @@
+- In Android 4.3, Chrome forces `preload` value to `none`.
+- On a cellular connection (2G, 3G, and 4G), Chrome forces `preload` value to
+  `metadata`.
+
```
Can you add a section on best practices? E.g. Prefer preload=metadata or even preload=none+poster if you have lots of resources on the same domain to avoid hitting the 6 connection HTTP connection cap which can hang loading of resources. This may also improve page speed if video is a secondary attraction on the page; i.e. you don't want to slow down page loading and scrolling by loading in clips that are unlikely to be used or where a brief delay is fine. In these cases prefer preload=none+poster.

beaufortfrancois replied:
Done.

dalecurtis commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,559 @@
+Before we dive into the [Cache API] and service workers, let's see how to
+manually buffer a video with MSE. The example below assumes that your web
+server supports HTTP Range requests. Note that this would be pretty similar to
+segment files.
```
Maybe note that there are entire middleware libraries dedicated to this; shaka, jwplayer, video.js, etc.

beaufortfrancois replied:
Done.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+<a href="#video_preload_attribute">Video preload attribute</a>
+      </td>
+      <td rowspan=3>
+Simple to use for an unique file hosted on a web server
```
nit: a unique

beaufortfrancois replied:
Done.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+
+## Video preload attribute
+
+If the video source is an unique file hosted on a web server, you may want to
```
nit: a unique

beaufortfrancois replied:
Done.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+Full control
+      </td>
+      <td>
+Complex error handling is website responsibility
```
nit: website's

beaufortfrancois replied:
Done.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+completely loaded and parsed (eg. `DOMContentLoaded` event) while the very
+different `load` event will be fired when resource has actually been fetched.
+
+<figure class="TODO">
```
This image is missing. On purpose? (https://pr-4744-dot-web-central.appspot.com/web/fundamentals/media/fast-playback-with-video-preload shows a broken link img thumbnail here)

beaufortfrancois replied:
Fixed.

beaufortfrancois replied:
Thanks!

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+
+Resource fetching will start only when initial HTML document has been
+completely loaded and parsed (eg. `DOMContentLoaded` event) while the very
+different `load` event will be fired when resource has actually been fetched.
```
The media element's `load` event is quite complex. It may be fired before the media is actually fetched (eg., during deferred load state, or prior to setting the src attribute to anything, and other cases. Note, there are also media element `loadstart` and `loadedmetadata` events, which may more clearly define what the app might be looking for.

johnpallett replied:
IIUC a key point here is that the preload attribute only works within the context of the current page, unlike link preload which can work on other pages (warming up the video before the user navigates to the page with the player). Please call that out specifically, it's an important reason why people might choose between the two approaches.

beaufortfrancois replied:
Sorry for the confusion there. I meant `window.onload` event not media element's `load` event. I've fixed it.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+    var bufferedSeconds = video.buffered.end(0) - video.buffered.start(0);
+    console.log(bufferedSeconds + ' seconds of video are ready to play!');
+  });
+</script>
```
What is the output of this example script? Note that just loading metadata may not buffer anything at all. Did you instead mean to await the `loadedmetadata` event somewhere in this example? Or perhaps observe when duration changes from NaN to something finite or +Infinity?

beaufortfrancois replied:
You're right. I've changed it to wait for `loadedmetadata` now.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
````diff
@@ -0,0 +1,571 @@
+</script>
+```
+
+Setting `preload` value to `auto` indicates that the browser should cache
````
This 'should' might be too strong wording here. The spec indicates: "Hints to the user agent that the user agent can put the user’s needs first without risk to the server, up to and including optimistically downloading the entire resource."

beaufortfrancois replied:
`s/should/may/`
Thanks!

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+attribute and set `preload` value to `none`. That way, you would avoid hitting
+the maxmium number of HTTP connections per host (6 in Chrome at the time of
+writing) which can hang loading of resources. Note that this may also improve
+page speed if video aren't part of your core user experience.
```
nit: videos

beaufortfrancois replied:
Fixed.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+allows you to force the browser to make a request for a resource without
+blocking the document's `onload` event. Resources loaded via `<link
+rel="preload">` are stored locally in the browser, and are effectively inert
+until they're explicitely referenced in the DOM, JavaScript, or CSS.
```
nit: explicitly

johnpallett replied:
clarify that this can be done on other pages. That's key to making the player page a fast load in a lot of cases.

beaufortfrancois replied:
@wolenetz Fixed.

@johnpallett I wouldn't recommend this. 

> Both prefetch ([RESOURCE-HINTS]) and preload declare a resource and its fetch properties, but differ in how and when the resource is fetched by the user agent: prefetch is an optional and low-priority fetch for a resource that might be used by a subsequent navigation; preload is a mandatory and high-priority fetch for a resource that is necessary for the current navigation. Developers should use each one accordingly to minimize resource contention and optimize load performance.

Source: https://w3c.github.io/preload/

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+video, audio, etc.). It should be used to warm up the browser cache for current
+sessions.
+
+<figure class="TODO">
```
ditto broken link

beaufortfrancois replied:
Fixed.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+<script>
+  // Later on, after some condition has been met, set video source to the
+  // preloaded video URL.
+  video.src = 'https://cdn.com/file.mp4';
```
This URL differs from the one used in preload. Does this work as you intend? :)

beaufortfrancois replied:
Thanks for the flag!
Fixed.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
````diff
@@ -0,0 +1,571 @@
+</script>
+```
+
+Note: I would recommend using this only for audio files or short video clips.
````
What are recommended sizes here?

beaufortfrancois replied:
It is better to be explicit. I've added `< 5MB`.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+
+<script>
+  // This function is automatically executed when the first segment of video is
+  // preloaded successfully. We may call it explicitely as well later on.
```
nit: explicitly

wolenetz replied:
Also, the text just before this example claims "We'll use MSE to feed the video element with those" (meaning the 3 segments), but this example shows only handling what to do once the very first segment's preload is complete.

wolenetz replied:
(Though it does include a TODO for the second segment, which helps, so maybe ignore my last comment :) )

beaufortfrancois replied:
Fixed nit.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
````diff
@@ -0,0 +1,571 @@
+```
+
+Warning: For cross-origin resources, make sure your CORS headers are set
+properly as opaque responses retrieved with `fetch(videoFileUrl, { mode:
````
nit: For clarity, would "because" be better here than "as"?

beaufortfrancois replied:
Fixed.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+Before we dive into the [Cache API] and service workers, let's see how to
+manually buffer a video with MSE. The example below assumes that your web
+server supports HTTP Range requests. Note that this would be pretty similar to
+segment files. It exists some middleware libraries such as [Google's Shaka
```
nit: fix awkward wording "It exists some middleware"...

beaufortfrancois replied:
Fixed.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+<script>
+  const mediaSource = new MediaSource();
+  video.src = URL.createObjectURL(mediaSource);
+  mediaSource.addEventListener('sourceopen', sourceOpen);
```
nit: once: true

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+    console.log(bufferedSeconds + ' seconds of video are ready to play!');
+
+    // Fetch the rest of the video when user starts playing video.
+    video.addEventListener('playing', fetchNextSegment, { once: true });
```
nit: s/fetchNextSegment/fetchRemainder/ with comment that a more complete example would ensure each append succeeds (including verifying buffered range expectations, handling any quota exceeded issues, and breaking "remainder" into chunks in size and for a stream bitrate that can be fetched without delaying playback) and so forth? Or is this too much? :)

beaufortfrancois replied:
`s/fetchNextSegment/fetchRemainder` applied.
Rest might be too much for this article focused on preload.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+### Pre-cache multiple first segments
+
+Now, what if I want to speculatively pre-load some media content without
+knowing which piece of media user will eventually pick. If user is on a web
```
nit: the user

beaufortfrancois replied:
Fixed.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+The 2-parts example below shows you how to pre-cache multiple first segments of
+video using the powerful and easy-to-use Cache API. Note that something similar
+can be achieved with IndexedDB as well. We're not using service workers yet as
+the Cache API is also accessible from the Window object.
```
Why is this window object important here? What am I missing?

beaufortfrancois replied:
It is important as developers might not know the Cache API (always used with Service Worker) can also be used without Service Worker, from the Window object.

That makes me think I should replace `caches` with `window.caches` in code snippets below.

beaufortfrancois replied:
Done.

wolenetz replied:
Ack. Thanks for clarifying.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
````diff
@@ -0,0 +1,571 @@
+
+```
+function onPlayButtonClick(videoFileUrl) {
+  video.load(); // Used to be able to play video later.
````
Why is load() required here, when setting the src attribute in l.458, below, triggers the media element's load algorithm?

beaufortfrancois replied:
See https://developers.google.com/web/updates/2017/06/play-request-was-interrupted and https://bugs.chromium.org/p/chromium/issues/detail?id=404161

wolenetz replied:
I see. Thanks for clarifying. Perhaps call out in a JS comment a link to at least https://developers.google.com/web/updates/2017/06/play-request-was-interrupted

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+  .then(data => {
+    const mediaSource = new MediaSource();
+    video.src = URL.createObjectURL(mediaSource);
+    mediaSource.addEventListener('sourceopen', sourceOpen);
```
nit: once: true

beaufortfrancois replied:
Fixed.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+    }
+
+    function updateEnd() {
+      video.play()
```
Playback could start prior to handling 'updateend' **if** either the element is autoplay, or you issue the play sometime after setting the src attribute (modulo the usual user gesture requirement caveats). For even faster playback start, this could help :)

beaufortfrancois replied:
Good call!
Fixed.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
````diff
@@ -0,0 +1,571 @@
+}
+```
+
+Warning: For cross-origin resources, make sure your CORS headers are set
````
I'm less clear around spec and implementation of these CORS requirements. For the specific case in the example with caching videoFileUrl's first segment using a range-based request, above, the "Response" that is cached is constructed from scratch and fed an arrayBuffer. Are all the CORS restrictions on media elements adhered to in that code path (would the range request's fetch response including { mode: 'no-cors' } preserve the disallowance of being fed to a media element even if that response's contents were extracted into an ArrayBuffer and used to construct and cache the new Response object?

beaufortfrancois replied:
It is impossible to read response body from an opaque response, hence why you need to set up CORS headers properly. @jakearchibald may want to confirm.

wolenetz replied:
See @jakearchibald comment https://github.com/google/WebFundamentals/pull/4744/files#r129577914

jakearchibald replied:
See my comment on a similar line earlier in the article

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+  return caches.match(request)
+  .then(response => {
+
+    // Fetch from network if it's not already in the cache.
```
You probably also want to cache the result of this fetch, because the same segment may be needed for MSE appendBuffer more than once over the lifetime of the cache.

beaufortfrancois replied:
I've added a comment as well as it is already shown in `fetchAndCache()`.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+</div>
+
+Have a look at the official [Sample Media App] and in particular its
+[ranged-response.js] file for a complete solution in how to handle Range
```
nit: s/in/for/

beaufortfrancois replied:
Fixed.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+    </tr>
+    <tr>
+      <td>
+MSE is not compatible
```
nit: maybe "MSE ignores preload attribute on media element because app is responsible for providing media to MSE" ?

dalecurtis replied:
+1, I think that'd be more intelligible.

johnpallett replied:
+1

beaufortfrancois replied:
Fixed.

dalecurtis commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+</video>
+
+Faster playback means more people watching your video. That's a known fact.
+I'll explore in this article different techniques you can use to accelerate
```
"In this article I'll explore different..." reads better I think.

beaufortfrancois replied:
Fixed.

dalecurtis commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+If your website contains plenty of video resources on the same domain, I would
+recommend you set the `preload` value to `metadata` or define the `poster`
+attribute and set `preload` value to `none`. That way, you would avoid hitting
+the maxmium number of HTTP connections per host (6 in Chrome at the time of
```
It's not just Chrome, per the HTTP 1.1 spec the max connections per host is 6. If you're using something like QUIC or HTTP/2 you can have as many as you want though.

beaufortfrancois replied:
I won't call Chrome specifically then. Thank you!

For info the HTTP 1.1 spec says:

```
   Clients that use persistent connections SHOULD limit the number of
   simultaneous connections that they maintain to a given server. A
   single-user client SHOULD NOT maintain more than 2 connections with
   any server or proxy. A proxy SHOULD use up to 2*N connections to
   another server or proxy, where N is the number of simultaneously
   active users. These guidelines are intended to improve HTTP response
   times and avoid congestion.
```

jpmedley commented:
@beaufortfrancois Is this ready to go? Has John looked at this?

johnpallett commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+
+### Support
+
+As it is not supported in every browser yet, you may want to detect its
```
clarify: Link preloading is not supported in every browser yet...

beaufortfrancois replied:
Fixed.

johnpallett commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+
+### Create Range responses with a Service Worker
+
+Now what if you fetched an entire video file and saved it in the Cache API. When
```
I'd add a quick intro explaining "If you use a Service Worker and the Cache API, you have complete control over when and how much data is cached."

johnpallett replied:
please have Jake Archibald review this section as well

beaufortfrancois replied:
I'll ping him.

johnpallett commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,571 @@
+Have a look at the official [Sample Media App] and in particular its
+[ranged-response.js] file for a complete solution in how to handle Range
+requests.
+
```
Above you mention that the site has to handle errors, do you want to add more detail on common errors and how to handle them here? It's OK to point to the sample app but it's worth listing a few things they'll have to manage

beaufortfrancois replied:
Paul Lewis does a good job at explaining this in the video below.

johnpallett commented:
I've left some comments. Please make sure the video is update (NOT YouTube) before publishing.

beaufortfrancois commented:
I've addressed feedback from @wolenetz, @dalecurtis, and @johnpallett at  https://github.com/google/WebFundamentals/pull/4744/commits/8facba7eddc163038efc7a7979566d7af4a42b22 and https://github.com/google/WebFundamentals/pull/4744/commits/deb68a6573199bebb660e730412976946965c91d

@jpmedley As soon as @jakearchibald has a chance to look at it, you may want to have a look at it.

jakearchibald commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,578 @@
+Compatible with MSE and segment files
+      </td>
+      <td>
+Should be used only for audio files or short video clips when fetching full
```
Why are audio files always ok?

beaufortfrancois replied:
I assumed audio files were short by default.
This is not always the case, I'll fix the wording.

jakearchibald commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,578 @@
+applied in Chrome:
+
+- When [Data Saver] is enabled, Chrome forces `preload` value to `none`.
+- In Android 4.3, Chrome forces `preload` value to `none`.
```
Not 4.2 or 4.4?

beaufortfrancois replied:
Jelly Bean: https://cs.chromium.org/chromium/src/content/renderer/render_view_impl.cc?type=cs&q=forcepreloadnone&sq=package:chromium&l=965

jakearchibald replied:
Hah wow. Might be worth saying "due to an android bug".

beaufortfrancois replied:
Done.

jakearchibald commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,578 @@
+
+### Tips
+
+If your website contains plenty of video resources on the same domain, I would
```
Maybe "many" instead of "plenty"? "Plenty" has a positive connotation.

beaufortfrancois replied:
Fixed.

jakearchibald commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,578 @@
+If your website contains plenty of video resources on the same domain, I would
+recommend you set the `preload` value to `metadata` or define the `poster`
+attribute and set `preload` value to `none`. That way, you would avoid hitting
+the maximum number of HTTP connections per host (6 according to the HTTP 1.1
```
Avoid mixing "host" and "domain"?

beaufortfrancois replied:
Fixed.
I now use "domain" for both.

jakearchibald commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,578 @@
+sessions.
+
+<figure>
+  <img src="/web/fundamentals/media/images/video-preload/link-preload.svg">
```
The video will start downloading *while* the page is downloading, no?

beaufortfrancois replied:
Thank you so much for pointing out this.
As I was playing with small html files, it always looked like the video started downloading after. I'll fix it.

jakearchibald commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,578 @@
+### Preload full video
+
+Here's how to preload a full video  on your website so that when your
+JavaScript asks to fetch video content, it is read from disk cache as the
```
I thought preload was a memory cache. I may be wrong.

jakearchibald replied:
(the spec doesn't require it either way)

beaufortfrancois replied:
Shall I simply omit "disk" from "disk cache"?

jakearchibald replied:
Yeah, that makes sense

beaufortfrancois replied:
Fixed.

jakearchibald commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
````diff
@@ -0,0 +1,578 @@
+
+```
+<link rel="preload" as="video" href="https://cdn.com/file_1.webm"
+    onload="preloadFinished()" onerror="preloadFailed()">
````
It's possible for these to run before these functions are defined right? Should we encouraging this pattern?

jakearchibald replied:
In fact, why do we need these callbacks at all? `fetch()` should just pick up the preloading item automatically.

beaufortfrancois replied:
I have an opinion but I'd like to hear @yoavweiss voice first ;)

yoavweiss replied:
That highly depends on what you're trying to do and if you need preload specific handling there. Otherwise, `fetch()` should pick up the preloaded item (again, assuming related issues are resolved).

beaufortfrancois replied:
After giving it more thoughts, I've decided to follow your advices and simplify lots of things.

beaufortfrancois replied:
https://github.com/google/WebFundamentals/pull/4744/commits/dfe6d4d55aff2fed7a47d506f91e74c64117af0f

jakearchibald commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,578 @@
+  }
+
+  function sourceOpen(event) {
+    URL.revokeObjectURL(video.src);
```
I'm not familiar with MSE, but this line is kinda surprising.

wolenetz replied:
MSE object URLs are not auto-revoking. To prevent bloat, explicitly revoking MSE object URLs when the MediaSource object has been successfully attached to the media element (as signalled by 'sourceopen' event) is a best practice. Only if you intend to re-use the object URL for some later reattachment might you want to retain it.

jakearchibald replied:
Seems weird to revoke it while it's still in use, but I guess it makes sense. I suppose the service worker cache API behaves in a kinda similar way.

beaufortfrancois replied:
For info, Chrome may support in the future `HTMLMediaElement.srcObject  = MediaSource` which would take care of this `revokeObjectURL` thing. See https://bugs.chromium.org/p/chromium/issues/detail?id=506273?

jakearchibald replied:
Yeah, that makes sense. Object URLs are a mess.

wolenetz replied:
Agreed. We (MSE spec) were originally going the route for autorevoking URLs IIRC (paralleling the FILE API at the time), but due to FILE API change before MSE got to REC (and srcObject not in MSE either), the push to REC reconciled the concerns by making MSE Object URLs non-auto-revoking. srcObject with MSE is indeed something that would alleviate this particular concern/pattern mess.

jakearchibald commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
````diff
@@ -0,0 +1,578 @@
+rel="preload">` has been fired.
+
+```
+<link rel="preload" as="video" href="https://cdn.com/file_1.webm"
````
Won't this result in a preload cache miss, since it's `fetch()` that does the fetching, not video?

beaufortfrancois replied:
I don't think so as the `as="video"` simply communicates the request priority (low in media) not who might use it IIUC.
@yoavweiss may confirm.

yoavweiss replied:
If `fetch()` does the fetching, we should have an `as=fetch` there. `as` communicates priority, CSP policy and resource type, and mismatches result in memcache misses, as @jakearchibald pointed out. Also, as `fetch()` preload matching is currently buggy, at least in Chrome, I'm not sure you want to encourage people to use that pattern right now.

beaufortfrancois replied:
/me understood everything the wrong way ;)
Thank you @yoavweiss and @jakearchibald. I'll fix this

![me](https://cdn.someecards.com/someecards/usercards/MjAxMy01YzE3MmE4ZWJkZmRlZDkw.png)

beaufortfrancois replied:
https://github.com/google/WebFundamentals/pull/4744/commits/40cd3a9134c0b50d9bbe229ada2518d349e3ef87

jakearchibald commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,578 @@
+
+Warning: For cross-origin resources, make sure your CORS headers are set
+properly because opaque responses retrieved with `fetch(videoFileUrl, { mode:
+'no-cors' })` are not allowed to feed any video or audio element.
```
You can use opaques responses to video elements, in fact it's the default way media elements request things.

You need CORS to make MSE work, but it requires array buffers, and you can't create an array buffer from an opaque response.

beaufortfrancois replied:
You're right, I'll be more specific and not blame opaque responses alone.

jakearchibald commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,578 @@
+properly because opaque responses retrieved with `fetch(videoFileUrl, { mode:
+'no-cors' })` are not allowed to feed any video or audio element.
+
+### Support
```
This section isn't needed if preload is used as an enhancement

beaufortfrancois replied:
This section has 2 goals:
- Explain how to detect `link rel=preload as=video` support. This is not documented anywhere on the internet. Only `link rel=preload`.
- Give developers the ability to measure perf gains when preload is and is not supported.

jakearchibald commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,578 @@
+  }
+
+  function fetchRemainder() {
+    fetch('file.webm', { headers: { range: 'bytes=567140-1196488' } })
```
`'bytes=567140-'` is fine if you're wanting the remainder

beaufortfrancois replied:
Argh. English translation issue there... My bad. I meant the rest of the video but more specifically the second segment as we would never download entirely the entire rest of the video.
I may want to simply rename this function to `fetchNextSegment()`. WDYT?

jakearchibald replied:
+1

wolenetz replied:
+1, with the additional change to the comment in the caller
From:
// Fetch the rest of the video when user starts playing the video
To:
// Fetch the next segment of video when user starts playing the video

And in fetchNextSegment(), add a TODO comment indicating that further segment fetching and appending is elided for simplicity?

beaufortfrancois replied:
Done.

jakearchibald commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
````diff
@@ -0,0 +1,578 @@
+</script>
+```
+
+### Battery awareness
````
This section might be overkill, maybe replace it with a considerations section & just link off to docs on the other APIs. Eg, you may also want to consider the `Save-Data` header.

beaufortfrancois replied:
As `navigator.getBattery()` is not documented anywhere in /web, I've thought we ought to show at least a snippet. Folllowing your feedback, I've reduced the size of the snippet, added a `Save-Data` header section, and moved it under a new "Considerations" section. Thank you!

jakearchibald commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,578 @@
+      return networkResponse;
+    });
+  });
+}
```
I wonder if these examples would be a lot easier to read as async functions. That presumes developers understand them though.

beaufortfrancois replied:
I've asked for it before and I was told to use them yet (sadly).

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
````diff
@@ -0,0 +1,578 @@
+</script>
+```
+
+Note: I would recommend using this only for audio files or short video clips (< 5MB).
````
Ditto @jakearchibald comment https://github.com/google/WebFundamentals/pull/4744/files#r129569028

beaufortfrancois replied:
Done.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,578 @@
+### Pre-cache multiple first segments
+
+Now, what if I want to speculatively pre-load some media content without
+knowing which piece of media the user will eventually pick. If user is on a web
```
nit (again :) ): the user

beaufortfrancois replied:
Fixed.

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,578 @@
+
+{% include "web/_shared/contributors/beaufortfrancois.html" %}
+
+Faster playback means more people watching your video. That's a known fact.
```
nit: Perhaps "faster playback start"? (To prevent possible confusion of faster playback with playbackRate)

beaufortfrancois replied:
You're right.
Fixed!

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,578 @@
+[ShakaPlayer] to ignore the response time as an indicator of network speed.
+
+<div class="video-wrapper">
+  <iframe class="devsite-embedded-youtube-video" data-video-id="f8EGZa32Mts"
```
Reminder: @johnpallett mentioned to make sure the video is updated (NOT YouTube) (https://github.com/google/WebFundamentals/pull/4744#issuecomment-315051645)

beaufortfrancois replied:
I've updated it at https://storage.googleapis.com/webfundamentals-assets/videos/video-preload-hero.webm ;)

wolenetz replied:
Oh - I was assuming @johnpallett meant all the videos, not just the hero one. I'll let him clarify though :)

jakearchibald commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,578 @@
+      .then(data => {
+
+        // Get start position from Range request header.
+        const pos = Number(/^bytes\=(\d+)\-$/g.exec(request.headers.get('range'))[1]);
```
This will fail with range requests like `bytes=500-600`

beaufortfrancois replied:
Argh!
Let me fix that.

beaufortfrancois replied:
I've simply removed the `$`. Thank you!

jakearchibald commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,578 @@
+        const slicedResponse = new Response(data.slice(pos), options);
+        slicedResponse.setHeaders('Content-Range': 'bytes ' + pos + '-' +
+            (data.size - 1) + '/' + data.size);
+        slicedResponse.setHeaders('X-From-Cache': 'true');
```
Might be worth mentioning why you're setting this header

jakearchibald replied:
Doh, you do mention it

jakearchibald commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
````diff
@@ -0,0 +1,578 @@
+```
+
+It is important to note that I used `response.blob()` to recreate this sliced
+response as this simply gives me a handle to the file while
````
Dunno if it's worth mentioning, but Firefox will re-copy the blob to disk, so it's pretty slow at this compared to Chrome.

https://github.com/whatwg/fetch/issues/569

beaufortfrancois replied:
Noted.
I'll point to this thread.

beaufortfrancois commented:
Thank you so much @jakearchibald! I've addressed most of your feedback at https://github.com/google/WebFundamentals/pull/4744/commits/7e2fe1adcdc1f36fe1ccc9049f9a8911bbb3ff40. I'm now waiting for @yoavweiss on https://github.com/google/WebFundamentals/pull/4744#pullrequestreview-52364950 and https://github.com/google/WebFundamentals/pull/4744#pullrequestreview-52363535

wolenetz commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -265,8 +265,9 @@ rel="preload">` has been fired.
 Warning: For cross-origin resources, make sure your CORS headers are set
-properly because opaque responses retrieved with `fetch(videoFileUrl, { mode:
-'no-cors' })` are not allowed to feed any video or audio element.
+properly. As we can't create an array buffer from an opaque response retrieved
```
There's still a second occurrence (later in the MSE portion of this doc) that is less explanatory (it doesn't mention array buffer, for example). Make the two warnings the same?

beaufortfrancois replied:
Done.

beaufortfrancois commented:
@jakearchibald If my latest patches look good to you, I'd like @jpmedley to have a final look.

jpmedley commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,569 @@
+manually buffer a video with MSE. The example below assumes that your web
+server supports HTTP Range requests but this would be pretty similar with
+segment files. Note that some middleware libraries such as [Google's Shaka
+Player], [JW Player], and [Video.js] are dedicated to handle this for you.
```
'dedicated to handle'

I think you mean 'built to handle'. If so, please change.

jpmedley commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,569 @@
+
+Setting the `preload` attribute to `auto` indicates that the browser may 
+cache enough such that the playback to end would be possible without 
+requiring a stop for further buffering.
```
Huh? Maybe: "the browser may cache enough data that complete playback is possible without requiring a stop for further buffering.

beaufortfrancois replied:
Done.

jpmedley commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,569 @@
+applied in Chrome:
+
+- When [Data Saver] is enabled, Chrome forces the `preload` value to `none`.
+- In Android 4.3, Chrome forces the `preload` value to `none` due to an [Android bug].
```
In both webview and Chrome for Android?

beaufortfrancois replied:
Webview that shipped in Android 4.3 doesn't have this enhancement. So it's only in Chrome.

jpmedley commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,569 @@
+manually buffer a video with MSE. The example below assumes that your web
+server supports HTTP Range requests but this would be pretty similar with
+file segments. Note that some middleware libraries such as [Google's Shaka
+Player], [JW Player], and [Video.js] are dedicated to handle this for you.
```
I think you mean 'built to handle'. If so, please change.

beaufortfrancois replied:
Done.

jpmedley commented:
@beaufortfrancois There are only a few things you need to change, though you might also review my [typo fixes](https://github.com/google/WebFundamentals/pull/4744/commits/e5c18e92b046057c0286e23dae7cb7d7e549de49)

ebidel commented on src/content/en/fundamentals/media/fast-playback-with-video-preload.md:
```diff
@@ -0,0 +1,569 @@
+description: Preload video and audio for faster playback.
+
+{# wf_published_on: 2017-07-27 #}
+{# wf_updated_on: 2017-07-27 #}
```
Please add `{# wf_blink_components: Blink>Media #}`

beaufortfrancois replied:
Done

yoavweiss commented:
Is this still waiting on me for something? (It's not clear from the comments)

jpmedley commented:
Yes. Please add {# wf_blink_components: Blink>Media #}

beaufortfrancois commented:
@yoavweiss We're good now I believe.

beaufortfrancois commented:
@petele Using `&lt;script>` instead of `<script>` is wrong. What can I do there to bypass these critical errors?

```

**ERRORS**
[`src/content/en/fundamentals/media/fast-playback-with-video-preload.md#L124`](https://github.com/Google/WebFundamentals/blame/99434dcea5807c03b7ec9a8a29189d2691dded89/src/content/en/fundamentals/media/fast-playback-with-video-preload.md#L124) - Unescaped script tag in markdown document, use &lt;script...
[`src/content/en/fundamentals/media/fast-playback-with-video-preload.md#L141`](https://github.com/Google/WebFundamentals/blame/99434dcea5807c03b7ec9a8a29189d2691dded89/src/content/en/fundamentals/media/fast-playback-with-video-preload.md#L141) - Unescaped script tag in markdown document, use &lt;script...
[`src/content/en/fundamentals/media/fast-playback-with-video-preload.md#L199`](https://github.com/Google/WebFundamentals/blame/99434dcea5807c03b7ec9a8a29189d2691dded89/src/content/en/fundamentals/media/fast-playback-with-video-preload.md#L199) - Unescaped script tag in markdown document, use &lt;script...
[`src/content/en/fundamentals/media/fast-playback-with-video-preload.md#L229`](https://github.com/Google/WebFundamentals/blame/99434dcea5807c03b7ec9a8a29189d2691dded89/src/content/en/fundamentals/media/fast-playback-with-video-preload.md#L229) - Unescaped script tag in markdown document, use &lt;script...
[`src/content/en/fundamentals/media/fast-playback-with-video-preload.md#L290`](https://github.com/Google/WebFundamentals/blame/99434dcea5807c03b7ec9a8a29189d2691dded89/src/content/en/fundamentals/media/fast-playback-with-video-preload.md#L290) - Unescaped script tag in markdown document, use &lt;script...

petele commented:
@beaufortfrancois - I've changed the behaviour of the script test (#4949) to warn when it finds a script tag. They're allowed in a limited number of scenarios, including code samples. In this case, your use is fine.

petele commented:
please remove the DO NOT MERGE tag when this is ready to go.

beaufortfrancois commented:
@petele We're ready to go.

beaufortfrancois commented:
I still get `[19:52:09] ERROR: src/content/en/fundamentals/media/fast-playback-with-video-preload.md An exception occurred in testMarkdown: ReferenceError: logWarn is not defined` error though

petele commented:
Merged and also fixed the logWarn issue. Will push live tomorrow.
